FEAT: Add Dockerized PyRIT with Jupyter Notebook Support

ErdemOzgen opened the pull request:
## Description
This PR addresses issue #767 by adding Docker configuration with Jupyter Notebook support for PyRIT.
fixes #767 
The implementation:
- Creates a Dockerfile using a Microsoft-provided base image (as recommended in the issue discussion)
- Configures a pre-configured environment with all necessary dependencies for PyRIT
- Installs PyRIT with all extras using `pip install pyrit[all]` to ensure complete functionality (including GCG)
- Integrates Jupyter Notebook support for interactive usage and tutorials
- Properly configures GPU support for PyTorch

This PR aims to make PyRIT more accessible and portable by allowing users to run it in a consistent, isolated environment without needing to modify their host system or create specialized conda environments.

Related issue: #767
cc: @romanlutz @bashirpartovi @rlundeen2

## Tests and Documentation
- Added documentation for building and running the Docker container
- Added sample Jupyter notebook demonstrating basic PyRIT functionality
- Tested the Docker build process and verified Jupyter notebook functionality
- Verified GPU support works properly with PyTorch integration

ErdemOzgen commented:
@microsoft-github-policy-service agree

romanlutz commented on docker/Dockerfile:
```diff
@@ -0,0 +1,61 @@
+# Base image from Microsoft Azure with Python 3.12
+FROM mcr.microsoft.com/azureml/minimal-py312-inference:20250310.v1
+
+LABEL maintainer="ErdemOzgen"
```
I'm a little unsure about this one tbh. I don't particularly care about the label here, but we usually don't put such things in the code. Needless to say, we acknowledge every contribution with every release and perhaps you could even write a blog post for the website blog if you want. Wdyt?

ErdemOzgen replied:
I completely agree with you. I had some doubts when I was initially writing it as well. I'll remove it right away.

romanlutz commented on docker/Dockerfile:
```diff
@@ -0,0 +1,61 @@
+FROM mcr.microsoft.com/azureml/minimal-py312-inference:20250310.v1
+
+LABEL maintainer="ErdemOzgen"
+LABEL description="Docker container for PyRIT with JupyterNotebook integration"
```
```suggestion
LABEL description="Docker container for PyRIT with Jupyter Notebook integration"
```
super nit

romanlutz commented on docker/Dockerfile:
```diff
@@ -0,0 +1,61 @@
+    seaborn
+
+# Install PyRIT from PyPI (the official way)
+RUN pip install --no-cache-dir pyrit[all]
```
```suggestion
RUN pip install --no-cache-dir pyrit[dev,all]
```
might sound counterintuitive but the `dev` dependencies include a few that are useful when running specific notebooks (like semantic-kernel)

romanlutz commented on docker/notebooks/example.ipynb:
I left another comment on notebooks (i.e., we should ideally include all that we have in the repo) but if we can only include one then it shouldn't be about GCG which is one of the less-used features (albeit a valuable one).

romanlutz reviewed (commented):
Thank you for the contribution @ErdemOzgen ! I left a few comments.

ErdemOzgen commented on docker/README.md:
```diff
@@ -0,0 +1,142 @@
+├── README.md
+├── data/                  # Persistent data storage
+├── docker-compose.yaml    # Docker Compose configuration
+├── notebooks/             # Jupyter notebooks directory
+│   └── example.ipynb      # Example PyRIT notebook
```
Added this functionality with adding this code in start.sh:

```bash
# Clone PyRIT repository if not already present
if [ ! -d "/app/PyRIT" ]; then
    echo "Cloning PyRIT repository..."
    git clone https://github.com/Azure/PyRIT
else
    echo "PyRIT repository already exists. Updating..."
    cd /app/PyRIT
    git pull
    cd /app
fi

# Copy doc folder to notebooks directory
echo "Copying documentation to notebooks directory..."
cp -r /app/PyRIT/doc/* /app/notebooks/
rm -rf /app/PyRIT
```

romanlutz replied:
I see! Make sure to add app/PyRIT/* to gitignore, though. Wouldn't want this to mess up version control

ErdemOzgen commented:
@romanlutz The PR is ready for your review now.

romanlutz commented on docker/notebooks/example.ipynb:
Given that you made the adjustment to add the other notebooks do we still need this one?

ErdemOzgen replied:
Done. good catch. Also update .gitignore with these lines:

```bash
# Ignore data directory in Docker folder
/docker/data/
/docker/data/*

# Ignore notebooks directory in Docker folder
/docker/notebooks/
/docker/notebooks/*
```

ErdemOzgen commented:
@romanlutz All done. Thank you for all your recommendations and reviews. It was a pleasure for me. I'm waiting for the maintainer team.

romanlutz commented:
@ErdemOzgen heads up @bashirpartovi just added a similar thing but for development in #787 and may push a few updates to unify these two ideas. It's probably best to have a single solution for both devs and users.

bashirpartovi reviewed (commented):
Left a few comments, please let me know if it makes sense

ErdemOzgen commented:
@bashirpartovi Thanks for feedback, I will rework this and get back to you.

ErdemOzgen commented:
@bashirpartovi All done—thanks a lot for your feedback! Definitely several good catches. Waiting for approval.

bashirpartovi commented:
@ErdemOzgen apologies for the delay, all the changes look great, I will manually pull down the docker tonight to test it just to make sure nothing missed my eyes :)

ErdemOzgen commented:
@bashirpartovi and @romanlutz, thank you both very much for your reviews and valuable feedback. I've incorporated your suggestions and updated the branch accordingly. It's now ready for your testing. Please let me know if you have any further comments or recommendations. It was truly a pleasure collaborating with you!

bashirpartovi commented on docker/README.md:
```diff
@@ -0,0 +1,186 @@
+      - "8888:8888"
+    volumes:
+      - ./notebooks:/app/notebooks
+      - ./data:/app/data
```
You need to mount `assets` folder as well. 
If you run `code/converters/4_image_converters.ipynb` notebook, you'll see that it complains it cannot find `../../../assets/pyrit_architecture.png`

bashirpartovi commented on docker/docker-compose.yaml:
```diff
@@ -0,0 +1,25 @@
+      - "8888:8888"
+    volumes:
+      - notebooks:/app/notebooks
+      - data:/app/data
```
You need to mount `assets` folder as well. 
If you run `code/converters/4_image_converters.ipynb` notebook, you'll see that it complains it cannot find `../../../assets/pyrit_architecture.png`

bashirpartovi commented on docker/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+RUN pip install --no-cache-dir pyrit[dev,all]
+
+# Create a directory for notebooks and data
+RUN mkdir -p /app/notebooks /app/data && \
```
You need to mount `assets` folder as well. 
If you run `code/converters/4_image_converters.ipynb` notebook, you'll see that it complains it cannot find `../../../assets/pyrit_architecture.png`

bashirpartovi commented on docker/docker-compose.yaml:
```diff
@@ -0,0 +1,25 @@
+      - notebooks:/app/notebooks
+      - data:/app/data
+    env_file:
+      - .env
```
You need a way to include PyRIT `.env` and `.env.local` (https://github.com/Azure/PyRIT/blob/main/.env_example ) in your image because Jupyter notebooks depend on those env vars.

bashirpartovi reviewed (commented):
@ErdemOzgen Looks great, I tried to test it last night and found some issues, added a few comments to fix :)

ErdemOzgen commented:
@bashirpartovi All done waiting for your comments.

bashirpartovi commented on docker/docker-compose.yaml:
```diff
@@ -0,0 +1,26 @@
+      - data:/app/data
+      - ../assets:/app/assets
+    env_file:
+      - .env
```
As I mentioned below, you can directly include the `.env` and `.env.local` into your container. 
```yaml
env_files:
   - ../.env
   - ../.env.local
   - .container.env.settings
```

bashirpartovi commented on docker/README.md:
```diff
@@ -0,0 +1,187 @@
+├── README.md            # This documentation file
+├── requirements.txt     # Python packages
+├── docker-compose.yaml  # Docker Compose configuration
+├── env.example          # Default and example env file (rename .env)
```
This probably needs to be updated based on my comments below

ErdemOzgen commented:
@bashirpartovi all done, waiting for your comments.

bashirpartovi commented on docker/README.md:
```diff
@@ -0,0 +1,189 @@
+      - ../assets:/app/assets
+    env_file:
+      - ../.env
+      - ../.env
```
it should be `../.env.local`

bashirpartovi commented on docker/docker-compose.yaml:
```diff
@@ -0,0 +1,28 @@
+      - ../assets:/app/assets
+    env_file:
+      - ../.env
+      - ../.env_local
```
It should be `../.env.local`

bashirpartovi commented on docker/docker-compose.yaml:
```diff
@@ -0,0 +1,28 @@
+    env_file:
+      - ../.env
+      - ../.env_local
+      - .container.env.settings
```
(suggestion)
I was thinking about this, since we have `.env.*` rule in the .gitignore file, it maybe a good idea to rename it as `.env.container.settings`. That way any changes users make to this file won't show up under the branch changes.

bashirpartovi reviewed (commented):
A couple of minor changes

ErdemOzgen commented:
@bashirpartovi all done. Waiting for review.

bashirpartovi commented:
Great job @ErdemOzgen , thanks for following up on all the feedbacks, I kicked off the checks so there may be other things that you may need to fix but approved from my side.
